Mathlib/CategoryTheory/Limits/Shapes/Reflexive.lean

```diff
@@ -148,6 +171,372 @@ instance (priority := 100) hasCoreflexiveEqualizers_of_hasEqualizers [HasEqualiz
 
+end CategoryTheory
+
+namespace CategoryTheory
+
+universe v v₂ u u₂
+
+namespace Limits
+
+/-- The type of objects for the diagram indexing reflexive (co)equalizers -/
+inductive WalkingReflexivePair : Type where
+  | zero
+  | one
+  deriving DecidableEq, Inhabited
+
+open WalkingReflexivePair
+
+/-- The type of morphisms for the diagram indexing reflexive (co)equalizers -/
+inductive WalkingReflexivePairHom : (WalkingReflexivePair → WalkingReflexivePair → Type)
+  | left : WalkingReflexivePairHom one zero
+  | right : WalkingReflexivePairHom one zero
+  | reflexion : WalkingReflexivePairHom zero one
+  | left_reflexion : WalkingReflexivePairHom one one
+  | right_reflexion : WalkingReflexivePairHom one one
+  | id (X : WalkingReflexivePair) : WalkingReflexivePairHom X X
+  deriving DecidableEq
+
+attribute [-simp, nolint simpNF] WalkingReflexivePairHom.id.sizeOf_spec
+
+open WalkingReflexivePairHom
+
+/-- Composition of morphisms in the diagram indexing reflexive (co)equalizers -/
+def WalkingReflexivePairHom.comp :
+    ∀ { X Y Z : WalkingReflexivePair } (_ : WalkingReflexivePairHom X Y)
+      (_ : WalkingReflexivePairHom Y Z), WalkingReflexivePairHom X Z
+  | _, _, _, id _, h => h
+  | _, _, _, h, id _ => h
+  | _, _, _, reflexion, left => id zero
+  | _, _, _, reflexion, right => id zero
+  | _, _, _, reflexion, right_reflexion => reflexion
+  | _, _, _, reflexion, left_reflexion => reflexion
+  | _, _, _, left, reflexion => left_reflexion
+  | _, _, _, right, reflexion => right_reflexion
+  | _, _, _, right_reflexion, right_reflexion => right_reflexion
+  | _, _, _, right_reflexion, left_reflexion => right_reflexion
+  | _, _, _, right_reflexion, right => right
+  | _, _, _, right_reflexion, left => right
+  | _, _, _, left_reflexion, left => left
+  | _, _, _, left_reflexion, right => left
+  | _, _, _, left_reflexion, right_reflexion => left_reflexion
+  | _, _, _, left_reflexion, left_reflexion => left_reflexion
+
+theorem WalkingReflexivePairHom.id_comp
+    {X Y : WalkingReflexivePair} (g : WalkingReflexivePairHom X Y) : comp (id X) g = g := by
+  cases g <;> rfl
+
+theorem WalkingReflexivePairHom.comp_id
+    {X Y : WalkingReflexivePair} (f : WalkingReflexivePairHom X Y) : comp f (id Y) = f := by
+  cases f <;> rfl
+
+theorem WalkingReflexivePairHom.assoc {X Y Z W : WalkingReflexivePair}
+    (f : WalkingReflexivePairHom X Y) (g: WalkingReflexivePairHom Y Z)
+    (h : WalkingReflexivePairHom Z W) : comp (comp f g) h = comp f (comp g h) := by
+  cases f <;> cases g <;> cases h <;> rfl
+
+instance walkingReflexivePairCategory : SmallCategory WalkingReflexivePair where
+  Hom := WalkingReflexivePairHom
+  id := id
+  comp := comp
+  comp_id := comp_id
+  id_comp := id_comp
+  assoc := assoc
+
+@[simp]
+theorem walkingReflexivePairHom_id (X : WalkingReflexivePair) :
+    WalkingReflexivePairHom.id X = 𝟙 X := by rfl
+
+@[simp]
+theorem walkingReflexivePairHom_left_comp_reflexion : left.comp reflexion = left_reflexion := rfl
```

Comment:
As much as possible, one the `Category` instance has been defined, we should use `≫` (`CategoryStruct.comp`) instead of the `comp` defined above. It would take the following form:
```lean
theorem walkingReflexivePairHom_left_comp_reflexion :
  (left ≫ reflexion : one ⟶ one) = left_reflexion := rfl```
```
(The annoying thing is that Lean cannot guess that `left` and `reflexion` are morphisms in a category, so that we need to help them a little bit.)
Actually, I am not convinced that this would be the best normal form, I would rather suggest keeping `left ≫ reflexion` as the normal form (and then `left_reflexion` is part of the internals of the implementation; by the way, `leftCompReflexion` may be a name that fits better our naming conventions!?)

Reply:
However, the following lemmas would be useful:
```lean
@[reassoc (attr := simp)]
theorem reflexion_comp_left : reflexion ≫ left = 𝟙 zero := rfl

@[reassoc (attr := simp)]
theorem reflexion_comp_right : reflexion ≫ right = 𝟙 zero := rfl
```
(With this combination of `reassoc` of `simp`, we should get reasonable normal forms for morphisms in that category.)